rfcs/20180827-api-names.md

```diff
@@ -0,0 +1,593 @@
+| **Sponsor**   | Martin Wicke (wicke@google.com)                      |
+| **Updated**   | 2018-08-27                                           |
+
+## Objective
+
+This document proposes organizing TensorFlow API symbols in corresponding logical subnamespaces. As TensorFlow library grows, it is important to structure namespaces in a clear way for easier discoverability and usability.
+
+We define **endpoint** as full name that can be used to access a TensorFlow
+symbol. For e.g.
+[name_scope](https://www.tensorflow.org/api_docs/python/tf/name_scope) can be
+accessed using either `tf.name_scope` or  `tf.keras.backend.name_scope`.
+Therefore, `name_scope` has 2 endpoints: `tf.name_scope` and
+`tf.keras.backend.name_scope`.
+
+At a high level we have the following goals:
+
+*   Add a few additional namespaces.
+*   Add additional endpoints for TensorFlow symbols in relevant namespaces.
+*   Remove some of the existing endpoints.
+
+## Motivation
+
+TensorFlow currently has over 2000 endpoints total including over 500 endpoints in the root namespace. As number of symbols grows, it is important to maintain a clear structure to aid discoverability.
+
+Certain API symbol placements could be improved:
+
+*   Some namespaces were created recently and might not contain all the
+    corresponding symbols. For e.g. `tf.math` namespace was added recently.
+    Symbols such as `tf.round` are not in `tf.math` namespace even though logically they belong in that namespace.
+*   Some symbols are included in the root namespace even though they are rarely
+    used (for e.g. `tf.zeta`).
+*   Some symbols currently start with a prefix that could really be replaced by
+    introducing a subnamespace (for e.g. `tf.string_strip` vs
+    `tf.strings.strip`, `tf.sparse_maximum` vs `tf.sparse.maximum`).
+*   Certain deep hierarchies seem redundant and could be flattened (for e.g.
+    `tf.saved_model.signature_constants.CLASSIFY_INPUTS` could be moved to
+    `tf.saved_model.CLASSIFY_INPUTS`).
+*   To keep clear structure and reduce duplication, we want to collect all layers, losses and metrics under the `tf.keras` namespace.
+
+In general, we want to balance flatness and browsability. Flat hierarchies allow
+for shorter endpoint names that are easy to remember (for e.g. `tf.add` vs
+`tf.math.add`). At the same time subnamespaces support easier browsability (for
+e.g. `tf.math` namespace would contain all math functions making it easier to discover available symbols).
+
+Furthermore, TensorFlow API has many users. Therefore, we should avoid removing endpoints if they are frequently used.
+
+## Design Proposal
+
+## Additional namespaces
+
+We plan to add the following additional namespaces:
+
+**tf.random** - will contain random sampling ops.
+**tf.keras.layers** - will contain all symbols that are currently under `tf.layers`. Note that signatures of these symbols will likely change to match layers under tf.keras.layers better.
```

Comment:
@martinwicke : Should this be rephrased (as are other references to `tf.layers`, since we do plan to remove the functional layer calls and also the variable-scope related wrapping of `tf.keras.Layers`?)

Reply:
Yes. I do think though that in addition to this rename RFC, we'll need another to publish the more detailed changes we're discussing.